### PR TITLE
WIP Circuit Breaker metrics

### PR DESCRIPTION
**Pull request by maciekciolek**
This PR is WIP over metrics for Circuit Breaker requested here https://github.com/akka/akka/issues/16617. It would be great to get feedback on the idea/approach.

I have already implemented raw events and time bucket streams. The goal is to implement stream of histograms using HdrHistogram - is it okey to add dependency for: https://github.com/HdrHistogram/HdrHistogram?

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-contrib/src/main/scala/akka/contrib/metrics/stage/CircuitBreakerEventsStage.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    var buffer: Buffer[Event] = _
+
+    override def preStart(): Unit = {
+      if (maxBuffer > 0) buffer = Buffer(maxBuffer, materializer)
+
+      breaker.onOpen(callback.invoke(Event.BreakerOpened(now)))
+      breaker.onHalfOpen(callback.invoke(Event.BreakerHalfOpened(now)))
+      breaker.onClose(callback.invoke(Event.BreakerClosed(now)))
+
+      breaker.onCallSuccess(e ⇒ callback.invoke(Event.CallSuccess(now, e)))
```
Comment (patriknw) [inline akka-contrib/src/main/scala/akka/contrib/metrics/stage/CircuitBreakerEventsStage.scala]: there will be quite a lot of overhead (generating actor messages via these async callbacks) so I think this call must be sampled by default (configurable)

**Review by patriknw — COMMENTED** [review]
Trying to look at this from 10,000 feet view first. I think it's useful but the question is where it should live?

We are trying to phase out akka-contrib (deprecated in 2.5) so that's not a good place.

Alpakka? Not really an integration with something external.

akka-stream-contrib, could work but I think we wanted to keep that without dependencies. Just small things built on top of Akka Streams.

[akka-management](https://github.com/akka/akka-management). It gives some insight to running system so that might make sense, especially if we also want to show these things in some dashboard?

WDYT?

**Comment by patriknw** [comment]
@maciekciolek see my previous comment. After talking to the team we agreed that it would be good to have this in https://github.com/akka/akka-management
Please open PR there, and no problem to add dependency to hdr there.

**Comment by nicoga97** [comment]
Hi! Were this metrics added? if yes in which package project ? Thanks!

**Pull request closed without merging**